packages/expo-image/src/useImage.ts

```diff
@@ -0,0 +1,48 @@
+'use client';
+
+import { useEffect, useState } from 'react';
+
+import { Image } from './Image';
+import type { ImageRef, ImageSource, UseImageHookOptions } from './Image.types';
+import { resolveSource } from './utils/resolveSources';
+
+/**
+ * A hook that loads an image from the given source and returns a reference
+ * to the native image instance, or `null` until the image is successfully loaded.
+ * @platform android
+ * @platform ios
+ */
+export function useImage(
+  source: ImageSource | string,
+  options: UseImageHookOptions = {}
+): ImageRef | null {
+  const resolvedSource = resolveSource(source) as ImageSource;
+  const [image, setImage] = useState<ImageRef | null>(null);
+
+  useEffect(() => {
+    async function loadImage() {
+      try {
+        const image = await Image.loadAsync(resolvedSource);
+        setImage(image);
```

Comment:
I was just thinking the other day, whether we're not overdoing it with hooks.

hooks come with a few issues as I see it.

- dependencies: does or does not a hook re-run when its parameter changes? Sometimes it makes sense, other times it does not. We need to document this. [Example](https://docs.expo.dev/versions/latest/sdk/font/#usefontsmap).
- reactivity (not relevant here): with permission hooks, we have recently discussed an issue: they give the feeling to consumers that whenever a permission changes, the hook will update its value. This is not universally true however. When you change permissions via the `settings` app, the hook won't update because there's no api that would inform the hook about the change.
- testing: I was recently fixing [an issue](https://github.com/expo/expo/pull/31143) that this code has as well. I tried to write a test for it but failed to repro it in test suite. I'm not sure how well the react-test-renderer mirrors the real behavior.

In this particular PR I think hook is fine but just wanted to point out.

Reply:
In general I agree with what you said. I'm not a big fan of hooks but it some cases they really simplify things and makes our API more ergonomic.
Also, `useVideoPlayer` and `useImage` can correctly handle when to release their native objects, out of the box. You can use an alternative, more imperative approach, but usually you may need to remember about calling the `release` function, otherwise the associated native objects will not free up the memory quickly. I just improved that for images (#31168) by providing the external memory footprint to the JS engine, but it's pretty hard for the audio/video player to find out how much virtual memory they use.

- dependencies: Yes, this needs to be clearly documented. Even without reading the docs, in `useImage` it seems quite intuitive that changing the source will update the image.
- reactivity: Good point – maybe we should detect when the app goes into foreground and update the permission hook?
- testing in RN is hard 😅

Reply:
> maybe we should detect when the app goes into foreground and update the permission hook?

that is an improvement but doesn't work when you have a split screen and change permissions without changing appState, or when you change permissions via adb or simctl (both tools support this). The hook api for permissions isn't the right abstraction unfortunately. Anyways, we'll discuss how to approach this as part of https://linear.app/expo/issue/ENG-12983